syncer/syncer.go

```diff
@@ -1404,16 +1446,48 @@ func (s *Syncer) syncDML(
 
 // Run starts running for sync, we should guarantee it can rerun when paused.
 func (s *Syncer) Run(ctx context.Context) (err error) {
-	tctx := s.tctx.WithContext(ctx)
+	runCtx, runCancel := context.WithCancel(context.Background())
+	defer runCancel()
+	tctx := s.tctx.WithContext(runCtx)
 
 	defer func() {
 		if s.done != nil {
 			close(s.done)
 		}
 	}()
 
+	go func() {
+		<-ctx.Done()
+		select {
+		case <-runCtx.Done():
+		default:
+			tctx.L().Info("received subtask's done")
+
+			s.waitTransactionLock.Lock()
+			if s.isTransactionEnd {
+				s.waitXIDJob.Store(int64(waitComplete))
+				s.jobWg.Wait()
+				tctx.L().Info("the last job is transaction end, done directly")
+				runCancel()
+				s.waitTransactionLock.Unlock()
+				return
+			}
+			s.waitTransactionLock.Unlock()
+
+			s.waitXIDJob.Store(int64(waiting))
```

Comment:
Can we switch these two lines?
Imagine that we run L982 to set `waitComplete` first, and then we set a `waiting` here. We must wait for one more transaction after that.

Reply:
L982 depends on CAS operation, so if the current status is not waiting, then it'll do nothing 🤔️

Reply:
> L982 depends on CAS operation, so if the current status is not waiting, then it'll do nothing 🤔️

I know. Then we will set `waitXIDJob` to `waiting` in L1477 and syncer will continue sending dml event to jobChan but we don't have to do that.

Reply:
> Can we switch these two lines?
> Imagine that we run L982 to set `waitComplete` first, and then we set a `waiting` here. We must wait for one more transaction after that.

SGTM, I will switch them in next commit.